### PR TITLE
OEP-55:ADR-5 Clarify Maintainer Interaction with Core Contributors

### PR DESCRIPTION
**Pull request by sarina**
This ADR clarifies the rights and responsibilities of Maintainers with respect to the Core Contributors.

Much of this interaction was previously inferred, but as it was not spelled out, there has been some confusion with respect to how much control Maintainers have over Core Contributors. In short, Maintainers do not control Core Contributors; CCs have passed a nomination process and have gained the support of the community to make good decisions in the repository. However, Maintainers are allowed to know who has CC rights on their repo, communicate repo standards and architectural vision to them, and initiate CC remove procedures if a CC is not following community standards and is unresponsive to feedback.

Relevant parts of OEP-54 Core Contributors & OEP-55 Maintainers are cited within the ADR to make clear that this ADR is not a new idea but rather cements the intentions of the original OEP authors.

Rendered: https://open-edx-proposals--566.org.readthedocs.build/en/566/processes/oep-0055/decisions/0005-managing-core-contributors.html

### CODE REVIEW

**Review by feanil — APPROVED** [review]

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (arbrandes) [inline oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: What is the process to remove a core contricutor?  It feels to me like giving maintainers the right to straight-up remove a contributor, without explicit recourse, might be too much.

Comment (feanil) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: This may require further clarity, my reading was that maintainers and CCs have already spoken directly about the reason for the removal and have agreed.  If the removal is for misconduct, then the maintainer would announce the removal along with misconduct citations.   The removal which would come in as an axim request would either have a link to the public discussion/citations or would be confirmed with both the CC and maintainer before we proceed.

Comment (feanil) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: Some of this is probably worth stating here.

Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: I meant they could initiate the CC removal process as described in https://docs.openedx.org/projects/openedx-proposals/en/latest/processes/oep-0054-core-contributors.html#removing-core-contributors - I will clarify this.

Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: Wait... I do say this. 

> Maintainers have the right to remove Core Contributors who are not appropriately following Open edX community standards, Core Contributor standards (the "3 Cs") or project expectations, and have also not been responsive to Maintainer feedback. OEP-54 specifies how to initiate this process

Could you elaborate on what's unclear here? I feel this paragraph very clearly (to me) describes what I meant: Maintainers should provide feedback to CCs, and if that is unheeded, they can follow OEP-54's provision to begin the removal procedure. No where do I say this is a unilateral privilege that maintainers can just boot a CC.

**Comment by antoviaque** [comment]
Looks great! :) Thanks for taking the time to clarify this, and for the balanced approach.

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (arbrandes) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:66]: No need for further clarification.  I must have glossed over the details at the end of the paragraph.  Thanks, and sorry for the noise!

**Review by arbrandes — APPROVED** [review]

**Review by felipemontoya — APPROVED** [review]
Looks good to me. Thanks for taking the time to clarify this.

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (hurtstotouchfire) [inline oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:53]: Does this really need to be stipulated? If maintainers no longer have admin access to repos isn't this just on Axim?

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: You bring up a good point that I want to spend a bit of time thinking about. I think feature flags can be a good thing (and a thing we want to encourage) in many situations, for example to enable gradual rollout or to gate a new feature for one Named Release before the new behavior becomes default.

Can you point to exactly where this line is coming from? I don't see it when I search for "toggle" in https://docs.openedx.org/projects/openedx-proposals/en/latest/processes/oep-0054-core-contributors.html

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:53]: It is important to explicitly state this, so that when a maintainer files an Axim request to add branch protection rules we can point to this ADR to say why it can't happen.

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:47]: I agree with your rewording, thank you. I will accept it after some more people look at this PR so they can see your prompt for further discussion. Personally, I don't think it's possible to define an "upper limit" because our repos vary so strongly in terms of scope/focus/complexity, `edx-platform` may always be an outlyer in terms of number of CCs simply because we need a plethora of CCs in order to cover all the different djangoapps and libraries contained within. I could see there being a reasonable case for an upper limit in some cases, for example a repo that had severe security implications on the rest of the platform. But in most cases, no, I don't think defining an upper limit makes sense.

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (brian-smith-tcril) [inline oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:59]: It might be nice to directly link to the CCs wiki page here for clarity. I know the maintainers how-to links to it, but it wasn't fully clear to me that I should click that link to find the answer to "Who is a CC on my repo?" as opposed to just "What should I do if I take over maintenance of a repo?"

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (hurtstotouchfire) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: It's here: https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Coding+Contributors+Materials#Committer-Rights-and-Responsibilities

I can't know what the intention was when that was written but I read this as more like "don't merge things that will break other operators' instances or cause surprising behavior", which does still seem valuable, but more like a responsibility than a right. I was looking for a better place to put that kind of information and I think that [Merge Guidelines for Coding Core Contributors](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3334635570/Merge+Guidelines+for+Coding+Core+Contributors) probably also needs a revamp as there is a lot of 2U specific stuff there. Might even be preferable to pull the contents of that page into an ADR on OEP-54 and hammer out the changes needed. Maintainer expectations for merge-readiness could just refer to the CC expectation.

I think the language in this line of the ADR is fine, we should just clean up the CC rights / responsibilities pages so this can refer to them.

**Review by hurtstotouchfire — COMMENTED** [review]
I think most of my concerns are about following up and laying out clear expectations for CC's & maintainers around merge readiness. Seems worth revising the CC rights and resonsibilities to generalize or remove any 2U-specific provisions and also make them more generic so we can then say they apply to maintainers as well.

Further discussion [here](https://github.com/openedx/open-edx-proposals/pull/566#discussion_r1516685379), but I don't think that needs to block this PR.

**Review by hurtstotouchfire — APPROVED** [review]

File: oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst
Comment (robrap) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: @hurtstotouchfire: I have also had thoughts on _merge readiness_, so let me know if you ticket or work on any related doc updates. Thanks.

Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: Ohhh yes. Goodness. Those pages definitely need a revamp; I'm embarrassed to say that's been on my todo list since January 😓

Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: @robrap @hurtstotouchfire - feel free to review those pages and leave a pile of comments. For context, Nimisha made those original pages in early 2020 and I believe the last meaningful update was by me in 2021. That is, pre-split.

Comment (sarina) [reply oeps/processes/oep-0055/decisions/0005-managing-core-contributors.rst:28]: @hurtstotouchfire - Feanil & I made a bunch of updates on these pages last week. Please feel free to provide a review, if you have time!

**Pull request merged**
